### PR TITLE
Add one-click increment/decrement buttons to the calendar

### LINKED ISSUE #4365: UX: add  Prev - Next Buttons next to Calendar
to get a good "feeling" for things that happen on your site, I'm sometimes switching from day to day, week to week or month to month instead of choosing the larger period.

It would be great to have a previous and next button beside the period selector to directly be able to choose the previous or following period (with the same duration: day,week,month..)

please see attachment.

further enhancement:
if you use this function very often, it would be awesome to be able to activate a performance mode, where period before and next would be preprocessed in the background so one could switch very fast...

### PR DESCRIPTION
**Pull request by michalkleiner**
Note — this PR replaces https://github.com/matomo-org/matomo/pull/21518 so that Matomo team can push it over the line with all the bells and whistles. Credit to @jimddavis.

Fixes #4365.

### Description:

Adds increment and decrement buttons with arrows to the date PeriodSelector component.
Allows one-click single period movement without having to open the calendar and select the previous/next day, week, month or year.

When dates range type is selected, the buttons are not shown.

For each period type, checks are done against PiwikMinDate and PiwikMaxDate, and the arrows are disabled if an increment or decrement would move beyond those dates. Min date is usually the date when the site was created (and can have data) and max date is today.

On small screens below 600px, the buttons are wider to increase the tap area.

The newly added `getWeek` method returns the week of year number. The algorithm was derived from https://www.w3resource.com/javascript-exercises/javascript-date-exercise-24.php and updated based on http://www.java2s.com/example/nodejs/date/get-the-iso-week-date-week-number.html for legibility.

period prev next preview

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Review by bx80 — APPROVED** [review]
I think The UI screenshot `enable_framed_pages_embed_whole_app.png` needs to be updated. 

A very minor UI issue: When hovering over the next period icon the style changes the chevron to grey, but also when the next period is unavailable the chevron is also shown in grey. Therefore if I'm hovering over the next icon I can't see whether the next period is available or not.

Other than that this works well for me and I can' t see any issues with the code :+1:

**Comment by michalkleiner** [comment]
@bx80 is it really grey on hover? It should be green, but I admit the difference is very subtle. In the image below it's hover on left, disabled on right.
period move buttons

I had a green background for the buttons as an alternative but that was out of scope for the 
ticket. I'll consult with Javi to see if we want to improve it visually in a follow up ticket.

2 extra screenshots updated, good catch!

**Comment by bx80** [comment]
I'm color blind so it might well be green not grey, either way I can't tell the difference :face_with_spiral_eyes: 
I've previously mentioned some of our colors are hard to differentiate for color blind users, but it's not seen as important. If it's easy enough for normal eyes to see then best just leave it as is :+1:

**Comment by michalkleiner** [comment]
I thought about it and created an alternative that has more prominent hover effect, but after double checking that implementation with Javi, for the time being, we will not use it as it would be a one off exception to the rest of the UI. I know we need to start somewhere but it would create an inconsistency in the UI, so I get their view as well.

Thanks for raising it though, it only confirms we need to do more in the A11Y space.

**Comment by jimddavis** [comment]
Thrilled you guys are adding this, but do have a question.  Is this the code you are working with?  https://github.com/matomo-org/matomo/tree/9866c649b791b537386e07d33c996e46fb4925c8/plugins/CoreHome/vue/src/PeriodSelector

If so, there are a number of bugs in it.  I am fairly new to working with GitHub so I may have made a mistake.  My most recent code, based on someone in your teams suggestions, is at  https://github.com/matomo-org/matomo/tree/9866c649b791b537386e07d33c996e46fb4925c8/plugins/CoreHome/vue/src/PeriodSelector

**Comment by jimddavis** [comment]
Thrilled you guys decided to include this feature. I have been re-implementing this in my dev environment with every significant release, and then deploying just those changes to my live server. It was a big pain, but I use the feature often.  Been holding off any any updates the past 6 months, and will wait till 5.1 is released before I upgrade from 4.16.

Not sure how many users will make use of it, but in my simple scenario, a quick glance at how things change day to day, or period to period, in various reports is very helpful.

Thanks for taking up the slack where my skill set was just not enough to get it done!

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue [hunk 1]
```diff
@@ -595,6 +625,119 @@ export default defineComponent({
...
+        case 'year':
+          atBoundary = year === boundaryDate.getFullYear();
+          break;
+        default:
+          break;
+      }
+
+      return !atBoundary;
+    },
+    getWeekNumber(date: Date) {
```
Comment (sgiehl) [inline plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: Should we maybe move those new methods to `plugins/CoreHome/vue/src/Periods/utilities.ts` for reusability?

**Review by sgiehl — APPROVED** [review]
Left a minor comment we could consider.

Otherwise this looks good and everything works as expected. Even when comparing dates, the compared periods are shifted correctly.

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue (hunk 1 above)
Comment (michalkleiner) [reply plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: Yep, can do that. I came across the Periods when I was looking if the methods existed somewhere already but then left it in the selector component. Will move it.

**Pull request merged**